Updating tcp

trygvis opened the pull request:
I tried to use the current tcp actor as it is, but couldn't figure out how to use it in a nice way and realized via #14 that ractor_cluster might contain better code so this PR updates the code with that code.

There is one main change is the addition of `Listener::on_connect`, used to spawn `Session` because the ractor_cluster code called the NodeSession stuff directly.

I have created a TCP example too: https://github.com/trygvis/ractor_actors/tree/tcp-example, will open a pull request for that too once this one is stable.

TODO:
* review if this is acceptable at all from @slawlor 
* Update comments, they're just copied from ractor_cluster right now. Some Object/Frame wording in particular needs improvement. I think some comments could be expanded on in general too.
*

trygvis committed:
fix: Typo

trygvis committed:
Updating net actor with code from ractor_cluster.

The code based are obviously based on each other so it was fairly small
number of changes.

The FameReceiver and SessionAcceptor traits are gone, they're replaced
by the new Session messages and callback that has to be configured on
the Listener.

trygvis committed:
net/tcp example

trygvis committed:
Updating comments

* Consistent use of "frame" instead of "object". Mostly because that is
what was there from before.
* Dropping "node", it's for the ractor_cluster code.

trygvis committed:
Better actor linking

Changing the way things are constructed so that the listener explicitly
spawns a Session for each connection. The Session will in turn call
spawn_handler that creates an actor and (optionally) links it to the
session. If linked to the session, it will also be supervised and torn
down on session closure, both explicit closure and failures.

trygvis commented:
I realized that the supervisor linking between the actors became less than ideal now so I'm doing a litte refactoring here.

trygvis committed:
Better actor linking

This works more like advertised where the listener, session and
reader/writer halves are linked.

This leads to a tightly coupled system, more so that really necessary.

trygvis commented:
After developing a version where both listener and sessions are linked I realize that it becomes too tightly coupled. I think the listener and session/reader/writer actors should be separate actors and the user should instead compose these as it want to.

The current Session also does the line packet format that I think should be refactored out of the networking layer, but that is perhaps a quest for another day.

trygvis committed:
Even better actor linking and organization

The consumer of the Listener actor will how spawn independent Session
actors and can choose their linking as it see fit.

Also, separating the listener and session modules. The contents of the
tcp module could be renamed to stream and the listener/session modules
would be independent and only depend on stream.

trygvis committed:
Adding watchdog actor

This actor will monitor registered actors and kill/stop them after a
specified unless it receives a ping message.

trygvis commented:
I've updated the code to a structure that I think is better, but would like some input. I've added a Watchdog actor that I have developed myself, it's at least useful to kill the tcp connections in the example :) Should show how the linking works, the MySession can be killed if the connection dies and the connection will be killed if MySession dies.

The example can run with and without the watchdog.

slawlor commented:
Just want to say I'm not ignoring this but I'm away for a week so it will take some time before I can look at this. Thanks!

trygvis commented:
@slawlor Great, no problem!

trygvis committed:
Watchdog fixes

slawlor commented on ractor_actors/src/net/tcp/listener.rs:
```diff
@@ -3,98 +3,88 @@
+/// The [Listener] supervises all the TCP [super::session::Session] actors and is responsible for
+/// logging connects and disconnects.
+pub struct Listener {
+    port: super::NetworkPort,
```
actors, ideally, shouldn't be constructed with arguments. That's why the actor startup routines have an Arguments type. This is because the actor only has access to `&self` and can't access owned instances of any of the provided values, nor can it mutate them if necessary. Ideally the "actor" type is just a stub struct, and the arguments & state contain all necessary owned logic.

slawlor commented on ractor_actors/src/net/tcp/listener.rs:
```diff
@@ -3,98 +3,88 @@
-}
-
-/// Arguments to startup a TcpListener
-pub struct ListenerStartupArgs<R>
```
i.e. here is why we have args.

slawlor commented on ractor_actors/src/net/tcp/listener.rs:
```diff
@@ -3,98 +3,88 @@
+        Self {
+            port,
+            encryption,
+            connection_handler: Arc::new(move |stream| Box::pin(connection_handler(stream))),
```
why do you want to take the v-table hit here, vs using a pure generic?

slawlor commented on ractor_actors/src/net/tcp/listener.rs:
```diff
@@ -3,98 +3,88 @@
 // LICENSE-MIT file in the root directory of this source tree.
 
-//! Tcp listener actor
+//! TCP Server to accept incoming sessions
```
in general most of the edits here seem to be a loss of flexibility and really mostly a no-op change? I'm not sure if this is a consequence of the clone of logic from ractor_cluster, but this version was written after and is more generic and flexible by design.

slawlor commented on ractor_actors/src/net/tcp/session.rs:
```diff
@@ -112,10 +156,15 @@ where
-        let (writer, _) =
-            Actor::spawn_linked(None, SessionWriter, write, myself.get_cell()).await?;
+        let (writer, _) = Actor::spawn_linked(
+            Some(format!("{}-rw", myself.get_name().unwrap_or_default())),
```
forcing named actors means a host of multiple actors of this type could just crash due to name clashes. naming is globally unique.

trygvis replied:
I didn't realize that they needed to be unique, I just used the name for logging myself.

slawlor commented on ractor_actors/examples/tcp.rs:
```diff
@@ -0,0 +1,205 @@
+            .await
+            .map_err(ActorProcessingErr::from)?;
+
+        if WATCHDOG.load(Ordering::SeqCst) {
```
this really should just be an arg to this actor, there's no reason it needs to be a static atomic bool

trygvis replied:
Yep, fixed in 9e2c53e2.

slawlor commented on ractor_actors/src/watchdog/mod.rs:
```diff
@@ -0,0 +1,75 @@
+use tokio::sync::OnceCell;
+
+pub enum TimeoutStrategy {
+    Kill(),
```
please don't make enums with arguments if they don't have arguments

```lang=rust
pub enum TimeoutStrategy {
    Kill,
    Stop,
}
```

slawlor commented on ractor_actors/src/watchdog/mod.rs:
```diff
@@ -0,0 +1,75 @@
+    }
+}
+
+async fn call<TReply, TMsgBuilder>(msg_builder: TMsgBuilder) -> Result<TReply, MessagingErr<()>>
```
are you dropping the message in the reply for a reason? Errors are supposed to send the error back to the caller, so the caller can retry/etc if they want. 

i.e. it should be MessagingErr<WatchdogMsg>

slawlor commented on ractor_actors/src/watchdog/mod.rs:
```diff
@@ -0,0 +1,75 @@
+use crate::watchdog::watchdog::WatchdogMsg;
```
Can we get some kind of docs on what this does? How it's working, what it's monitoring/etc.

slawlor replied:
Biggest question, how does the watchdog "watch" actors? What is it doing to ping them or check they're alive?

slawlor commented:
So just from reviewing the TCP changes, they seems pretty much like non-functional changes and go away from the more standard best-practices. I haven't had a ton of time to backport best practices to ractor_cluster, but I would much prefer we start with making an example tcp server/client with the existing framework. From your original post, I don't see what exactly was blocking about the original way it's written, and I'd like to focus on that before reverting logic to older patterns.

trygvis commented:
I'm a bit busy myself now, but I'll get back to this soon.

trygvis committed:
Updating after feedback

trygvis committed:
Updating after feedback

trygvis committed:
Adding docs on the watchdog module

trygvis committed:
Rolling back some unnecessary changes

trygvis committed:
Rolling back some unnecessary changes

Updating examples to new API

trygvis committed:
Comments and nits

trygvis committed:
ci: Enable on all branches

trygvis committed:
Clippy fixes

trygvis committed:
Fixing Clippy warning

trygvis committed:
Adding missing async_trait

trygvis committed:
cargo fmt --all

slawlor commented on .github/workflows/ci.yaml:
```diff
@@ -1,8 +1,6 @@
 name: CI
 on:
   push:
-    branches:
```
uhh why?

trygvis replied:
As my branch became longer-living than expected, I wanted to run the build on my branch too. Happy to roll back if you want to keep it in main only.

trygvis replied:
I guess also because the actions doesn't run automatically on this PR. But I also often have several branches that I'd like to have checked before creating the PR or while testing random stuff.

slawlor replied:
Please add this back so merges to `main` will trigger CI runs.

trygvis replied:
I'm a bit surprised it is not building on main, it is building on my branch: https://github.com/trygvis/ractor_actors/actions/runs/14497493422/job/40668831660.

I've put it back, but for later it would be nice to get both branches to build.

slawlor commented on ractor_actors/src/net/tcp/mod.rs:
```diff
@@ -3,27 +3,14 @@
-
-pub mod listener;
-pub mod session;
+//! TCP server and session actors which [Frame] messages
```
does this pass `cargo` docs generation?

trygvis replied:
Yes, the comment seems to go in the module overview page. I noticed that the sentence isn't quite complete so I'll fix that.

slawlor commented on ractor_actors/src/net/tcp/session.rs:
```diff
@@ -149,21 +169,19 @@ where
             }
-            Self::Msg::FrameReady(msg) => {
+            FrameReady(msg) => {
                 tracing::debug!(
```
ah we should probably drop this to trace level, this will be VERY noisy

trygvis replied:
Yes. I think it is a bit too much to have them there at all, if someone want to trace the bytes they could just do it themselves. Though it is sometimes nice to see exactly the bytes that are going in/out on the wire. Perhaps it could be put in a trait that the users could mix in?

I'm updating to `trace` for now.

slawlor commented on ractor_actors/src/net/tcp/session.rs:
```diff
@@ -188,6 +206,7 @@ where
                 } else {
                     tracing::error!("TCP Session received a child panic from an unknown child actor ({}) - '{}'", actor.get_id(), panic_msg);
                 }
+                myself.stop_children(Some("session_stop_panic".to_string()));
```
this is done automatically on teardown, is there a reason you need to do this manually?

trygvis replied:
Yes, at least I did when working on the other branch I wanted to merge.

I think the problem was that I had two actors that I wanted to be linked with each other, but it doesn't seem possible with linking that these are co-dependent.

In Erlang this would be solved with a third supervisor that used the "one for all" strategy. Not sure what the best pattern would be here.

slawlor commented on ractor_actors/examples/tcp.rs:
```diff
@@ -0,0 +1,241 @@
+
+struct MyServer;
+
+struct MyServerState {}
```
if it's an empty struct, please don't assign struct args to it. You just can do `struct MyServerState;` and it'll be compiled down to a nothing. 

However if the state is actually empty, the unit struct (`()`) is more idiomatic.

slawlor commented on ractor_actors/examples/tcp.rs:
```diff
@@ -0,0 +1,241 @@
+impl SessionAcceptor for MyServerSocketAcceptor {
+    async fn new_session(&self, stream: NetworkStream) -> Result<(), ActorProcessingErr> {
+        tracing::info!("New connection: {}", stream.peer_addr());
+        Actor::spawn(
```
these are spawned unlinked to the parent? They should be linked children of the server, so when the server shuts down all the children are terminated (i.e. network streams)

trygvis replied:
Hm, perhaps. From the TCP/IP side, there is no problem keeping the client socket alive even if the listening socket has been closed. 

Anyway, it is up to the application given that it is implementing the `SessionAcceptor` trait and this example should be as ractor idiomatic as possible so I'll update it.

slawlor commented on ractor_actors/examples/tcp.rs:
```diff
@@ -0,0 +1,241 @@
+    async fn frame_ready(&self, f: Frame) -> Result<(), ActorProcessingErr> {
+        self.session
+            .cast(MySessionMsg::FrameReady(f))
+            .map_err(ActorProcessingErr::from)
```
you don't need the conversion, `?` will automatically do it for you.

```
self.session.cast(MySessionMsg::FrameReady(f))?;
Ok(())
```

trygvis replied:
Ah, sweet!

slawlor commented on ractor_actors/examples/tcp.rs:
```diff
@@ -0,0 +1,241 @@
+}
+
+#[cfg_attr(feature = "async-trait", async_trait::async_trait)]
+impl Actor for MySession {
```
OK I think there's a bit of a misunderstanding in the pattern I intended here.

I believe you shouldn't need this `MySession` actor at all. By defining the frame receiver trait to do the higher-level logic you're doing inside the `handle()` method below, you can get rid of this actor all together.

The `TcpSession` will call your frame receiver caller for you, and you can do whatever conversion/decoding is necessary there.

slawlor replied:
I guess the only reason might be the integration to the watchdog bit here. But I'm not really clear _why_ you're compounding them together. Is it just to check if the session had activity in a certain time period?

I believe you could do this on the frame receiver, and for every frame ping the watchdog, instead of having a custom actor to do it.

trygvis replied:
The watchdog is to ensure client interactivity on a higher level than the TCP timeouts. I wanted to include it in the example as a way to show off how actors are automatically killed (and thus resources cleaned up).

I guess this example might be outdated after my last refactoring, I think I have to take a step back and look at everything again, in particular with the FrameReceiver change idea you mentioned.

slawlor commented on ractor_actors/examples/tcp.rs:
```diff
@@ -0,0 +1,241 @@
+
+#[cfg_attr(feature = "async-trait", async_trait::async_trait)]
+impl FrameReceiver for MyFrameReceiver {
+    async fn frame_ready(&self, f: Frame) -> Result<(), ActorProcessingErr> {
```
in retrospect it probably makes sense for this to take `&mut self` and have an owned copy for each `TcpSession`. That way it can act as the "state" of the tcp session and you can put whatever logic here for every received frame.

Additionally maybe it should receive the TcpSession handle as an argument, so if it needs to reply, it can.

trygvis replied:
That might be a good idea, but I think I have to try it out a bit first.

If I look at this from an Erlang point of view, it feels a little bit strange that the behavior isn't inside an actor. The FrameReceiver has no way of getting messages and is quite limited in what it can do. I think you would end up with a applicaion-level session actor anyway quite fast.

slawlor reviewed (changes requested):
OK this looks a lot more reasonable, and keeps with the intent of the original design. I added some other comments, and I think the example TCP could be cleaned up a bit, but it's pretty close IMHO so thanks for all the work!

trygvis committed:
Better docs

trygvis committed:
Reducing tracing level

trygvis committed:
Rust improvements

trygvis commented:
Thanks for the review @slawlor. Sorry for the delay in replying, I missed the notification email.

I have been using the ractor framework for a while now and I really like how close it is to Erlang compared to the others I've tried. I'm no expert in Rust (but know my way around in many other languages) so I get trapped with silly Rust mistakes sometimes. Thanks for pointing those out :) I've done work in both Erlang and Elixir, but Java/Go/C++ has been my main languages so I'm very happy to see such a strong actor framework in Rust.

I've been working on more examples, trying to work out and document how the framework is indented to be used. This `updating-tcp` branch was supposed to be a small update PR, but then I went down the wrong path. Some of the changes in this PR is meant to support stuff from the other branch, like the watchdog stuff: https://github.com/trygvis/ractor_actors/tree/tcp-example. Sorry if this is a bit confusing.

The `[tcp-example](https://github.com/trygvis/ractor_actors/tree/tcp-example)` branch is rebased on top of this, but the code might not be ideal given the latest comments about using separate actors for the session (`MySession`) so I'll have to review that separately. The diff is obviously quite a bit big until this PR is merged. 

The examples branch does quite a bit more refactoring so there is more to talk about there, but let's get this one done first.

trygvis commented:
Oh, and the Git history is kinda useless with all the smaller commits and fixups that I've done so either squash it all or let me fix up the history.

trygvis committed:
Rust improvements

trygvis commented:
I also wanted to add that I see the watchdog as an experiment for now. I'm using it in its current form in another project, so it is useful and works. However, I started with an interface of a module and some public methods. Its internals are hidden, even if it uses actors internally. This is the way OTP does it with a lot of its code. I personally think the interface becomes quite clear and easy to use, but it is up to you.

slawlor reviewed (approved):
Thanks for taking the time through all the diff reviews, I think this is in a solid enough state to merge for now.

slawlor commented:
Is there anything more you want to add here? 

Oh and last thing, can you run a cargo semver check after bumping the release version? I'm curious to see if it thinks we have semver breaks due to trait changes at all. I'm going to guess yes, but i'm not sure from looking at it raw.

slawlor commented on ractor_actors/src/net/tcp/mod.rs:
```diff
@@ -90,36 +89,3 @@ pub enum IncomingEncryptionMode {
-    async fn new_session(&self, session: NetworkStream) -> Result<(), ActorProcessingErr>;
-}
-
-pub use listener::*;
```
yeah this specifically I think will be a semver break. Not the end of the world, but can we re-export the types at the module root for `tcp`?

trygvis replied:
Sure. I guess it's expected that people only want to import the tcp module as today, so I guess `session::*` should be exported too?

trygvis replied:
I've re-exported both listener and session symbols to the tcp module, `cargo semver-checks check-release` shows no breakage now.

slawlor commented:
Also a couple very high-level comments about the watchdog framework. 

1. I like the api and with the back & forth comments and updated docs, I understand it's usage better
2. I think it might be worth exploring in the future ways to "scale" it up. In a single process, if it were used with a lot of subscribers being "watched", it'll probably start to choke since it's effectively a single-threaded process. It'll still handle a lot of load likely, but at some point it might need to be sharded into something like a factory. But not worth it at this stage for sure.
3. It'd be great if the `ping` call were synchronous. The async await, which is fallible, means an actor could fail on the watchdog bit which is kind of odd. Especially if the watchdog like crashes for whatever reason.

trygvis commented:
> Also a couple very high-level comments about the watchdog framework.
> 
>     1. I like the api and with the back & forth comments and updated docs, I understand it's usage better
> 
>     2. I think it might be worth exploring in the future ways to "scale" it up. In a single process, if it were used with a lot of subscribers being "watched", it'll probably start to choke since it's effectively a single-threaded process. It'll still handle a lot of load likely, but at some point it might need to be sharded into something like a factory. But not worth it at this stage for sure.

This is not going to be a very accurate watchdog, in particular if there are CPU bound tasks. If a single-threaded runtime is used any CPU bound task will keep the watchdog from processing its timeouts. If Ractor ever gets priority on its messages (either as a public or internal API), or a separate thread/thread pool that can be improved.

>     3. It'd be great if the `ping` call were synchronous. The async await, which is fallible, means an actor could fail on the watchdog bit which is kind of odd. Especially if the watchdog like crashes for whatever reason.

Synchronous as in `call(|reply| WatchdogMsg::Ping(actor, reply)).await`?

trygvis committed:
Symbol re-export fixup

No semver breakages reported with `cargo semver-checks check-release`
with this change.

trygvis committed:
watchdog: Making the ping call synchronous

trygvis committed:
watchdog: Note on CPU starvation

trygvis commented:
I think I'm happy to merge it now. The API breakage is fixed, however my next branch is likely going to mess up that! 

Seeing the code coverage going down is motivating to make a test case for the TCP code though ;)

trygvis committed:
ci: Putting back branch specification
